project/dexter-vs/Dexter/Common/Core/Utils/DexterSolutionManager.cs

```diff
@@ -1,34 +1,97 @@
 ﻿using Microsoft.VisualStudio.Shell.Interop;
 using System;
 using System.Collections.Generic;
+using System.Runtime.InteropServices;
 using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
+using System.IO;
 using System.Diagnostics;
 using Microsoft.VisualStudio;
 
 namespace Dexter.Common.Utils
 {
     /// <summary>
+    /// EventArgs for SourceFileChanged event
+    /// </summary>
+    public class SourceFileEventArgs : EventArgs
+    {
+        private bool isAdded;
+        private IList<string> filePaths;
+
+        /// <summary>
+        /// Creates an instance of SourceFileEventArgs
+        /// </summary>
+        /// <param name="filePaths">Path list of changed files</param>
+        /// <param name="isAdded">Whether files are added or not</param>
+        public SourceFileEventArgs(IList<string> filePaths, bool isAdded)
+        {
+            this.filePaths = filePaths;
+            this.isAdded = isAdded;
+        }
+
+        /// <summary>
+        /// Path list of changed files
+        /// </summary>
+        public IList<string> FilePaths
+        {
+            get { return filePaths; }
+        }
+
+        /// <summary>
+        /// Whether files are added or not
+        /// </summary>
+        public bool IsAdded
+        {
+            get { return isAdded; }
+        }   
+    }
+
+    /// <summary>
+    /// Provides interfaces for events of solution and project 
+    /// </summary>
+    public interface IDexterSolutionManager
+    {
+        /// <summary>
+        /// Provides file list changed event
+        /// </summary>
+        event EventHandler<SourceFileEventArgs> SourceFilesChanged;
+    }
+
+    /// <summary>
     /// Manages VS soultions and packages information/events
     /// </summary>
-    public class DexterSolutionManager : IVsSolutionEvents
+    public class DexterSolutionManager : IVsSolutionEvents, IDexterSolutionManager
     {
-        static DexterSolutionManager instace;
+        static DexterSolutionManager instance;
+        IDexterHierarchyService hierarchyService;
+
+        public event EventHandler<SourceFileEventArgs> SourceFilesChanged;
+
+        public DexterSolutionManager(IDexterHierarchyService hierarchyService)
+        {
+            this.hierarchyService = hierarchyService;
+        }
 
         static public DexterSolutionManager Instance
         {
             get
             {
-                if (instace == null)
```

Comment:
I have removed singletons that are not really needed. However, sometimes a singleton is required. For example, in the previous commit, you added the following code, where the ServiceProvider is also singleton. All VS services that can be imported via MEF are also singletons.

```
        [Import]
        public SVsServiceProvider ServiceProvider;
```